tools/nl/loadtest/loadtest.py

```diff
@@ -72,9 +95,79 @@ def load_test(query_file, output_file):
   print('')
 
 
+# Asynchronous function to run a query and add result of the run to results
+async def run_query(results, key, session, query):
+  try:
+    print(f'Query {key}: {query}')
+    response = await session.get(_URL + query)
+    # Check the response status code
+    if response.status == 200:
+      # parse the json response and add to the result
+      data = await response.json()
+      print(f'Query {key} completed successfully')
+      debug_timing = data.get('debug', {}).get('timing', {})
+      charts = data.get('charts', [])
+      chart_types = list(set([c['type'] for c in charts]))
+      results[key] = {
+          _RESULT_KEY_STATUS: 'EMPTY' if len(charts) == 0 else 'SUCCESS',
+          _RESULT_KEY_QUERY: query,
+          _RESULT_KEY_NUM_CHARTS: str(len(data['charts'])),
+          _RESULT_KEY_CHART_TYPES: chart_types,
+          _RESULT_KEY_NL_TIME: debug_timing.get('getNlResult', 0),
+          _RESULT_KEY_TILE_TIME: debug_timing.get('getTileResults', 0),
+          _RESULT_KEY_TOTAL_TIME: debug_timing.get('total', 0)
+      }
+    else:
+      print(f'Query {key} failed with status code:', response.status_code)
+      results[key] = {
+          _RESULT_KEY_STATUS: 'FAILURE',
+          _RESULT_KEY_QUERY: query,
+      }
+  except Exception as e:
+    print(f'Error running query {key}: {str(e)}')
+    results[key] = {
+        _RESULT_KEY_STATUS: 'FAILURE',
+        _RESULT_KEY_QUERY: query,
+    }
+
+
+# Runs asynchronous load test where all the queries are requested in parallel
+async def run_async_load_test(query_file, output_file):
+  results = {}
+  result_keys = []
+  async with aiohttp.ClientSession() as session:
+    with open(query_file) as inf:
+      tasks = []
```

Comment:
Should we control the parallelism?

Reply:
Do we not want to issue all the queries in parallel? I've updated to using concurrent.futures where we could set how many threads to have, but for the parallel mode, I use as many threads as there are queries

Reply:
So when you ran this what was the QPS?

The number of queries seems too few.  What I think might be helpful as a throughput test is, to run this + say https://github.com/datacommonsorg/website/blob/ab8cc12df92b283b523b63d3aedfc623517e2e2d/static/js/apps/explore_landing/topics.json queries in a loop as a sustained load for ~30 mins at a fixed parallelism of say `100`.  Then we look at the QPS.  Does that make sense?